### PR TITLE
Add generic reduction domains and Triton row-vector scheduling

### PR DESCRIPTION
**Pull request by whjthu**
`pytest` output:
```
pytest -n 16 -x
====================================================================================== test session starts =======================================================================================
platform linux -- Python 3.12.3, pytest-9.0.3, pluggy-1.6.0                                
rootdir: /home/haojie/ntbackend/.merge-ready/reduction-domain-triton-schedule                                                                                                                     
configfile: pyproject.toml                                                                       
plugins: xdist-3.8.0, anyio-4.13.0, cov-7.1.0
16 workers [387 items]                                                                           
...................................................................................................................................................s...................................... [ 48%] 
.......................................................................................................................................................................................... [ 96%]
s..............                                                                                                                                                                            [100%]
=========================================================================== 385 passed, 2 skipped in 529.47s (0:08:49) ==========================================================================
```

## 背景

当前 SSA 编译链路对 reduction 的调度主要依赖启发式判断，缺少对归约域、并行域和结果使用关系的结构化分析。这带来两个问题：

1. Triton 可能使用 block mask 访问 scalar pointer，导致生成代码无法编译。
2. Softmax、LayerNorm 等包含归约与广播的数据流无法稳定生成高效的 row-vector kernel，性能明显低于旧版生成路径。

本 PR 从 SSA 类型、归约轴和 use-def 链中提取通用的 Reduction Domain，并据此完成调度。

### 通用 Reduction Domain

新增不可变的 `ReductionDomain`，用于描述：

- reduction operand、result 和 `sum/max/min` 类型；
- 规范化后的 reduction axis；
- 输入、结果及 parallel domain shape；
- reduction extent；
- 同域 reduction 及其广播 consumer；
- row-vector 调度的合法性和拒绝原因。

分析同时覆盖：

- 正轴与负轴归一化；
- 多个同域 reduction；
- reduction 到算术、数学、条件、广播和 store 的数据流；
- `scf.if`、`scf.for`、`scf.yield` 和 loop-carried value；
- `mem.store` 与 `mem.atomic_add` 等 effect sink；
- 不同输出域、未知布局和不兼容 consumer 的 fail-closed 处理。

### Triton Row-Vector 调度

当 Reduction Domain 满足约束时，Triton 使用 row-vector 调度：

- 一个 Triton program 对应一个 parallel domain 实例；
- `tl.arange()` 覆盖 reduction axis；
- block pointer、value 和 predicate 保持相同向量域；
- 使用 `tl.sum`、`tl.max` 或 `tl.min` 完成归约；
- reduced output 使用 scalar store；
- 广播后的完整 shape output 使用 vector store；
- 支持任意单轴归约，包括 axis 0、末轴和 rank >= 3 的中间轴。

不满足条件时进入明确的 scalar fallback，无法保证语义时直接报错，不生成可疑代码。

### 调优接入

复用现有 `LaunchPlan` 和 AutoTuner，不引入第二套调优系统：

- 默认 `num_warps` 候选为 `4、8、1`；
- `num_stages=1`；
- 用户显式配置仍具有最高优先级；
- `max_num_configs=1` 使用安全的 4-warps 配置；
- winner 继续由现有运行时 benchmark 和缓存机制管理。

### 后端隔离

Row-vector value semantics 仅由 Triton 后端消费。CUDA 和 TileLang 保持原有 lowering 行为，后续 PR 再做更新。

## 正确性覆盖

新增测试集中覆盖以下高价值场景：

- Reduction Domain 的 axis、shape 和同域多归约分析；
- `sum/max/min` 生成；
- Softmax 和 LayerNorm 形式的通用 SSA 数据流；
- axis 0、末轴及 rank-3 中间轴归约；
- 非二次幂 reduction extent；
- 动态 shape、stride 和非连续输入；
- 不兼容广播及不同输出域的 fail-closed；
- tuning candidate 传播和显式配置优先级。

没有为每个内部 helper 增加独立测试，测试主要保持在 IR 契约、生成源码和公共执行路径三个层次。

## 测试结果

三个后端分别执行完整并行测试：

| 默认后端 | 结果 |
|---|---:|
| Triton | 385 passed, 2 skipped |
| CUDA | 385 passed, 2 skipped |
| TileLang | 385 passed, 2 skipped |

两个 skip 均为已有的多 GPU 测试，当前测试服务器只有一张 GPU：

- AOT multi-device test
- Triton multi-context reload test

## 性能结果

测试平台为 NVIDIA L20。基准为历史 master：

`c9ebd4950a185beed8d4c1db9ff4a1fd133934ae`

覆盖：

- Softmax：`4096 × {128, 256, 512, 1024, 2048, 4096}`
- LayerNorm：`4096 × {128, 256, 512, 1024, 2048, 4096}`
- Held-out width：`127、781、1127、4097`

共 12 个主性能点和 8 个 held-out 点，全部数值正确。

下表中的比值为：

`本 PR 延迟 / 历史 master 延迟`

因此小于 1 表示本 PR 更快。

| 指标 | 结果 |
|---|---:|
| 12 个主性能点延迟比几何平均 | 0.9659× |
| 相对历史 master 的整体提升 | 约 3.4% |
| 最差单点延迟比 | 1.0029× |
| 最差单点回归 | 约 0.29% |

所有主测试点均不超过历史 master 10%，整体性能优于历史 master。

## 改动范围

本 PR 只完成通用 Reduction Domain 和 Triton row-vector 物化。CUDA、TileLang 的并行归约物化与后端专属性能优化将通过后续独立 PR 完成。

### CODE REVIEW

File: src/ninetoothed/backends/emitters/ssa.py [hunk 1]
```diff
@@ -804,6 +926,27 @@ def _emit_value(name: str, ctx: _EmitContext) -> str:
     local = _local_symbol(name, ctx)
 
     if op.opcode.startswith("reduce."):
+        if ctx.vector_program:
```
Comment (voltjia) [inline src/ninetoothed/backends/emitters/ssa.py]: [P1] 仅对 schedule 选中的 reduction 执行向量化 lowering

`ctx.vector_program` 是 Program 级开关，但 `analyze_reductions()` 会把 scalar reduction 排除在 `reduction_schedule["reductions"]` 之外。因此，只要同一个 Program 中还有一个被调度的 tensor reduction，这里就会把 scalar reduction 一并向量化。

最小复现：

```python
import torch
import ninetoothed.language as ntl
from ninetoothed import Symbol, Tensor, make

W = Symbol("W", constexpr=True)

def arrangement(x, out, W=W):
    return x.tile((1, W)), out.tile((1,))

def application(x, out):
    row = ntl.sum(x, axis=1)
    scalar = ntl.sum(x[0, 0], axis=0)
    out = row + scalar

kernel = make(
    arrangement,
    application,
    (Tensor(2), Tensor(1)),
    backend="triton",
    max_num_configs=1,
)
x = torch.tensor([[1., 2., 3.], [4., 5., 6.]], device="cuda")
out = torch.empty(2, device="cuda")
kernel(x, out, W=3)
print(out)  # 实际: [9., 27.]；预期: [7., 19.]
```

在 `c12bf33` 上，schedule 只列出 row sum，但生成源码包含两个 `tl.sum`；实际结果正好等于 scalar 值被重复 `W` 次。

建议仅在当前 reduction result 位于 `schedule["reductions"]` 中时进入这条路径，或者让混合场景回退；同时增加验证数值结果的回归测试。

File: src/ninetoothed/backends/emitters/ssa.py [hunk 2]
```diff
@@ -753,7 +875,7 @@ def _emit_operation(op: ssa.Operation, ctx: _EmitContext) -> None:
         else:
             mask = _store_mask(
                 ctx.target,
-                ctx.mask_expr,
+                _mask_for_axes(_value_axes(tensor, ctx), ctx),
```
Comment (voltjia) [inline src/ninetoothed/backends/emitters/ssa.py:885]: [P1] 根据 source-store 的实际坐标选择 mask

上面的 source-store 分支仍使用 Program 级的 `ctx.mask_expr`。对于 row-vector reduction 的输出，渲染后的地址和 reduction 结果都是 scalar，但 `ctx.mask_expr` 是 block-valued 的 `offsets < W` mask。

最小复现：

```python
import torch
import ninetoothed.language as ntl
from ninetoothed import Symbol, Tensor, make

W = Symbol("W", constexpr=True)

def arrangement(x, out, W=W):
    return x.tile((1, W)), out.tile((1,))

def application(x, out):
    out.source[out.offsets(0)] = ntl.sum(x, axis=1)

kernel = make(
    arrangement,
    application,
    (Tensor(2), Tensor(1)),
    backend="triton",
    max_num_configs=1,
)
x = torch.ones((2, 3), device="cuda")
out = torch.empty(2, device="cuda")
kernel(x, out, W=3)
```

该例可以完成 lowering，但在 Triton JIT 编译阶段失败：

```text
Mask argument cannot be block type if pointer argument is not a block
```

建议像当前修改后的普通 store 和 load 路径一样，根据 `rendered` 坐标决定 base mask，并增加 row reduction 写入 source-store 的回归测试。

File: src/ninetoothed/compiler/runtime.py [hunk 3]
```diff
@@ -1183,6 +1192,17 @@ def _validate_tensor_contract(spec, value, expected_device):
             f"expected {source_ndim}."
         )
 
+    try:
+        expected_shape = tuple(int(dim) for dim in spec.attrs["source_shape"])
```
Comment (voltjia) [inline src/ninetoothed/compiler/runtime.py]: [P2] 按维度校验 source shape 中的固定部分

当前代码在同一个 `try` 中转换完整的 `source_shape` tuple。只要其中一个维度是 symbolic，`expected_shape` 就会整体变成 `None`，从而关闭所有维度的校验。

最小复现（直接命中 runtime contract）：

```python
from types import SimpleNamespace
import torch
from ninetoothed.compiler.runtime import _validate_tensor_contract

spec = SimpleNamespace(
    name="x",
    ndim=2,
    attrs={"source_ndim": 2, "source_shape": ("rows", "127")},
)
x = torch.empty((3, 64), device="cuda")

_validate_tensor_contract(spec, x, None)
print("accepted")  # 当前会执行到这里；预期应抛出 TypeError
```

这里第 0 维允许动态，但固定第 1 维必须等于 127。建议逐维比较，只跳过真正为 symbolic 的维度。回归测试应覆盖 `Tensor(shape=(None, 127))`，并覆盖 JIT、AOT 和 reload 三条路径。

File: src/ninetoothed/compiler/reductions.py [hunk 4]
```diff
@@ -0,0 +1,727 @@
...
+
+    for operation, scope in scoped_operations:
+        if not operation.opcode.startswith("reduce."):
+            continue
+
+        if operation.operands:
+            operand_type = value_types.get(operation.operands[0])
+
+            if operand_type is not None and operand_type.kind == "scalar":
+                _validate_scalar_fallback(operation)
```
Comment (voltjia) [inline src/ninetoothed/compiler/reductions.py:89]: [P3] 在进入 scalar fallback 前校验 reduction opcode

这条分支会在 `_domain()` 执行 `sum`/`max`/`min` 白名单检查之前返回。

最小复现：

```python
from ninetoothed.compiler.reductions import analyze_reductions
from ninetoothed.ir import ssa

scalar = ssa.Type(kind="scalar", dtype="float32")
x = ssa.Value(name="x", type=scalar)
result = ssa.Value(name="result", type=scalar)
program = ssa.Program(
    kind="repro",
    inputs=(x,),
    outputs=(result,),
    blocks=(ssa.Block(operations=(
        ssa.Operation(
            opcode="reduce.product",
            operands=("x",),
            results=(result,),
            attrs={"axis": 0},
        ),
    )),),
)

ssa.verify_program(program)
print(analyze_reductions(program))
# 当前: reduction_domains=(), reduction_schedule={"mode": "none"}
# 预期: 报告不支持 reduce.product
```

建议在 scalar/tensor 分流之前统一检查 opcode，或者把相同的白名单检查加入 `_validate_scalar_fallback()`；同时增加不支持的 scalar `reduce.*` analyzer 回归测试。

**Review by voltjia — APPROVED** [review]
